0004-remove-spandata.md

```diff
@@ -0,0 +1,43 @@
+# Remove SpanData
+
+Remove and replace SpanData by adding span start and finish options.
+
+## Motivation
+
+SpanData represents an immutable span object, creating a fairly large API for all of the fields (12 to be exact). It exposes what feels like an SDK concern and implementation detail to the API surface. As a user, this is another API I need to learn how to use. As an implementer, it’s a new data type that needs to be supported. The primary motivation for removing SpanData revolves around the desire to reduce the size of the tracing API.
+
+## Explanation
+
+SpanData's primary use case comes from the need to construct and report out of band spans, meaning that you're creating "custom" spans for an operation you don't own. A good example of this is a program that takes in structured logs that contain correlation IDs and a duration (e.g. from splunk) and [converts them](https://github.com/lightstep/splunktospan/blob/master/splunktospan/span.py#L43) to spans for your tracing system. [Another example](https://github.com/lightstep/haproxy_log2span/blob/master/lib/lib.go#L292) is running a sidecar on an HAProxy machine, tailing the request logs, and creating spans. SpanData supports the out of band reporting case, whereas you can’t with the current Span API as you cannot set the start and end timestamp, or add any tags at span creation that the sampler might need.
+
+I'd like to propose getting rid of SpanData and `tracer.recordSpanData()` and replacing it by allowing `tracer.startSpan()` and `span.finish()` to accept start and end timestamp options. This reduces the API surface, consolidating on a single span type. Options would meet the requirements for out of band reporting.
```

Comment:
I think what you are proposing is not enough:
* You may need a way to pass the "spanId" if in the logs you have the SpanId generated (I may be wrong but this is what I think you may also need) - we can deffer this but worth documented.
* You may need a way to pass events with a timestamp as well.

Reply:
I struggle with the idea of allowing users to set an explicit span ID but I think I understand the use case. For example, you have a sidecar that's accepting serialized spans that already have a span and trace ID, say from a legacy tracing system, that you want to convert to opentelem spans. I think if you're doing that you might just want to take a direct dependency on the sdk's span type and translate to that? If you only have a trace ID, you can set the it by passing in a span context with that trace id. I'll add a note that you may want to include a `withSpanID`.

I mention `withEvents()` in the internal details, does that cover the second bullet point? Alternatively `span.addEvent` could accept an optional timestamp.